src/style.rs

```diff
@@ -4,19 +4,81 @@ use crate::format::{BinaryBytes, DecimalBytes, FormattedDuration, HumanBytes, Hu
 use crate::state::ProgressState;
 use crate::utils::{expand_template, pad_str};
 use std::borrow::Cow;
+use std::collections::HashMap;
 
 #[cfg(feature = "improved_unicode")]
 use unicode_segmentation::UnicodeSegmentation;
 
+pub type Format = fn(&ProgressState) -> String;
+type FormatMapInner = HashMap<&'static str, Format>;
+
+#[derive(Clone)]
+pub(crate) struct FormatMap(FormatMapInner);
+
+impl Default for FormatMap {
```

Comment:
I think we should leave the built-in formatters outside of the format map, the map should be used only to augment/replace the default set of keys. In `expand_template()`, let's cache whether the map is empty and only do lookups if it isn't.

Reply:
Alright, that makes sense. However, I'm not really sure how you would do this in `expand_template` as it is right now. Should `expand_template` also take a reference to the hash map of custom keys as well? Also, how would it know if the key was not found without changing the `format_state` function? Currently, it just falls back to returning an empty string and just using that in the template. Perhaps this could be changed to return an `Option<String>`?

To me, it seems easier to just do all of that inside of the closure in `format_state`. It doesn't really seem necessary to cache whether the map is empty, as it seems very cheap to check for that by just calling `.is_empty()`: https://godbolt.org/z/M5fEPjWE4

Reply:
Ah, sorry, I guess I mean sort of like it is now, in the closure passed to `expand_template()`.

Reply:
Ok, will do. One implication of that approach is that the default keys cannot be overridden, because they are always matched first. Perhaps this should be mentioned in the documentation?

Reply:
The fact that default template keys are not overridable is now mentioned in the docstring of `with_key`.

Reply:
This change has been reverted, as the default keys are now overridable.